crates/shared/src/price_estimation/factory.rs

```diff
@@ -0,0 +1,292 @@
+                        self.components.http_factory.create(),
+                        self.args
+                            .balancer_sor_url
+                            .clone()
+                            .context("balancer SOR url not specified")?,
+                        self.network.chain_id,
+                    )?),
+                    self.rate_limiter(kind),
+                    self.components.gas_price.clone(),
+                )),
+            };
+
+        Ok(InstrumentedPriceEstimator::new(estimator, kind.name()))
+    }
+
+    fn get_estimator(&mut self, kind: PriceEstimatorType) -> Result<Arc<dyn PriceEstimating>> {
+        if let Some(existing) = self.estimators.get(&kind) {
+            return Ok(existing.clone());
+        }
+
+        let estimator = Arc::new(self.create_estimator(kind)?);
+        self.estimators.insert(kind, estimator.clone());
+        Ok(estimator)
+    }
+
+    fn get_estimators(
+        &mut self,
+        kinds: &[PriceEstimatorType],
+    ) -> Result<Vec<(String, Arc<dyn PriceEstimating>)>> {
+        kinds
+            .iter()
+            .copied()
+            .map(|kind| Ok((kind.name(), self.get_estimator(kind)?)))
+            .collect()
+    }
+
+    fn sanitized(&self, estimator: Box<dyn PriceEstimating>) -> SanitizedPriceEstimator {
+        SanitizedPriceEstimator::new(
+            estimator,
+            self.network.native_token,
+            self.components.bad_token_detector.clone(),
+        )
+    }
+
+    pub fn price_estimator(
+        &mut self,
+        kinds: &[PriceEstimatorType],
+    ) -> Result<Arc<dyn PriceEstimating>> {
+        let estimators = self.get_estimators(kinds)?;
+        Ok(Arc::new(self.sanitized(Box::new(
+            CompetitionPriceEstimator::new(estimators),
+        ))))
+    }
+
+    pub fn fast_price_estimator(
+        &mut self,
+        kinds: &[PriceEstimatorType],
+        fast_price_estimation_results_required: NonZeroUsize,
+    ) -> Result<Arc<dyn PriceEstimating>> {
+        let estimators = self.get_estimators(kinds)?;
+        Ok(Arc::new(self.sanitized(Box::new(
+            RacingCompetitionPriceEstimator::new(
+                estimators,
+                fast_price_estimation_results_required,
+            ),
+        ))))
+    }
+
+    pub fn native_price_estimator(
+        &mut self,
+        kinds: &[PriceEstimatorType],
+    ) -> Result<Arc<dyn NativePriceEstimating>> {
+        let estimator = Arc::new(CachingNativePriceEstimator::new(
+            Box::new(NativePriceEstimator::new(
+                Arc::new(
+                    self.sanitized(Box::new(CompetitionPriceEstimator::new(
+                        kinds
+                            .iter()
+                            .copied()
+                            .map(|kind| {
+                                Ok((
+                                    kind.name(),
+                                    Arc::new(self.create_estimator(kind)?)
```

Comment:
Note that we use `create_estimator` instead of `get_estimator` here. I kept it in, as this is what we currently do in the orderbook startup and didn't want to change any logic in this PR.